### PR TITLE
Refactor marshalling info parser to split overall logic from ordering

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Split our marshalling attribute parser into separate classes for each attribute and category and make the parser itself only handle executing the various different stages of marshalling info parsing. All logic for parsing a given attribute is handled now by the separate classes. Our special handling for types like arrays or SafeHandle-derived types is now also handled separately.

This PR allows the JS marshaller to reuse the core of our parsing and opt-in to more of the logic as they see fit.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
I've also updated some more portions of the JavaScript generators marshalling parsing logic while I was working on this area. @pavelsavara can you take a look at the JS side when you have time?

**Comment by pavelsavara** [comment]
I'm ooo till next week, perhaps @maraf could have look.

File: src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs [hunk 1]
```diff
@@ -4,185 +4,192 @@
...
+                    return new JSSimpleTypeInfo(KnownManagedType.IntPtr)
+                    {
+                        Syntax = SyntaxFactory.IdentifierName("nint")
+                    };
+                case { SpecialType: SpecialType.System_DateTime }:
+                    return new JSSimpleTypeInfo(KnownManagedType.DateTime)
+                    {
+                        Syntax = SyntaxFactory.ParseTypeName(fullTypeName.Trim())
+                    };
+                case ITypeSymbol when fullTypeName == "global.System.DateTimeOffset":
```
Comment (maraf) [inline src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs:75]:
```suggestion
                case ITypeSymbol when fullTypeName == "global::System.DateTimeOffset":
```

File: src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs [hunk 2]
```diff
@@ -4,185 +4,192 @@
...
+                    return new JSSimpleTypeInfo(KnownManagedType.DateTimeOffset)
+                    {
+                        Syntax = SyntaxFactory.ParseTypeName(fullTypeName.Trim())
+                    };
+                case ITypeSymbol when fullTypeName == "global.System.Exception":
```
Comment (maraf) [inline src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs:80]:
```suggestion
                case ITypeSymbol when fullTypeName == "global::System.Exception":
```

**Review by maraf — COMMENTED** [review]
This is probably the reason CI fails for JS interop

**Review by maraf — COMMENTED** [review]
JS generator changes looks good to me.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/StringMarshallingInfoProvider.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                    {
+                        countInfo = useSiteInfo.CountInfo;
+                    }
+                    return CustomMarshallingInfoHelper.CreateNativeMarshallingInfo(
+                        type,
+                        _defaultMarshallingInfo.StringMarshallingCustomType,
+                        _stringMarshallingCustomAttribute,
+                        useSiteAttributes,
+                        marshallingInfoCallback,
+                        true,
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/StringMarshallingInfoProvider.cs:49]: Perhaps we should use a named argument here. Also, where it is used with `false`.

**Review by AaronRobinsonMSFT — COMMENTED** [review]
Thanks for the solid doc on the new interfaces - very helpful.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingInfoParser.cs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    /// A provider of TypePositionInfo instances for other elements in the current signature, such as parameters or return values.
+    /// </summary>
+    public interface IElementInfoProvider
+    {
+        /// <summary>
+        /// Get the name for a given index.
+        /// </summary>
+        /// <param name="paramIndex">The index</param>
+        /// <returns>The name associated with the provided index, or <see cref="string.Empty"/> if the index does not correspond to an element.</returns>
+        string FindNameForParamIndex(int paramIndex);
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingInfoParser.cs:107]: Callers can accept any valid string. The returned value is used for diagnostics and for passing to `TryGetInfoForElementName`, which uses the try pattern. As a result, I don't think the `Try` pattern for `FindNameForParamIndex` is required.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ArrayMarshallingInfoParser.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Interop
+{
+    public sealed class ArrayMarshallingInfoProvider : ITypeBasedMarshallingInfoProvider
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ArrayMarshallingInfoParser.cs:13]: That depends on how plug-and-play we want these types to be. By having them public, developers of other marshalling source generators can pick and choose which ones are relevant for their marshalling models, like how the Javascript generator only uses a few of the provider implementations.

[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ArrayMarshallingInfoParser.cs]: Makes sense. Public is fine, but then we should add some comments.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingInfoParser.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingInfoParser.cs:107]: Makes sense.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**